zcash_client_sqlite/src/testing.rs

```diff
@@ -0,0 +1,814 @@
+            create_proposed_transaction,
+            input_selection::{GreedyInputSelectorError, InputSelector, Proposal},
+            propose_transfer, spend,
+        },
+    },
+    keys::{sapling, UnifiedFullViewingKey, UnifiedSpendingKey},
+    proto::compact_formats::{
+        self as compact, CompactBlock, CompactSaplingOutput, CompactSaplingSpend, CompactTx,
+    },
+    wallet::OvkPolicy,
+    zip321,
+};
+use zcash_note_encryption::Domain;
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::{BlockHeight, Network, NetworkUpgrade, Parameters},
+    legacy::TransparentAddress,
+    memo::MemoBytes,
+    sapling::{
+        note_encryption::{sapling_note_encryption, SaplingDomain},
+        util::generate_random_rseed,
+        value::NoteValue,
+        Note, Nullifier, PaymentAddress,
+    },
+    transaction::{
+        components::{amount::BalanceError, Amount},
+        fees::FeeRule,
+        TxId,
+    },
+    zip32::{sapling::DiversifiableFullViewingKey, DiversifierIndex},
+};
+
+#[cfg(feature = "transparent-inputs")]
+use zcash_client_backend::data_api::wallet::shield_transparent_funds;
+#[cfg(feature = "transparent-inputs")]
+use zcash_primitives::{
+    legacy, legacy::keys::IncomingViewingKey, transaction::components::amount::NonNegativeAmount,
+};
+
+use crate::{
+    chain::init::init_cache_database,
+    error::SqliteClientError,
+    wallet::{
+        commitment_tree,
+        init::{init_accounts_table, init_wallet_db},
+        sapling::tests::test_prover,
+    },
+    AccountId, ReceivedNoteId, WalletDb,
+};
+
+use super::BlockDb;
+
+#[cfg(feature = "unstable")]
+use crate::{
+    chain::{init::init_blockmeta_db, BlockMeta},
+    FsBlockDb,
+};
+/// A builder for a `zcash_client_sqlite` test.
+pub(crate) struct TestBuilder<Cache> {
+    cache: Cache,
+    seed: Option<SecretVec<u8>>,
+    with_test_account: bool,
+}
+
+impl TestBuilder<()> {
+    /// Constructs a new test.
+    pub(crate) fn new() -> Self {
+        TestBuilder {
+            cache: (),
+            seed: None,
+            with_test_account: false,
+        }
+    }
+
+    /// Adds a [`BlockDb`] cache to the test.
+    pub(crate) fn with_block_cache(self) -> TestBuilder<BlockCache> {
+        TestBuilder {
+            cache: BlockCache::new(),
+            seed: self.seed,
+            with_test_account: self.with_test_account,
+        }
+    }
+
+    /// Adds a [`FsBlockDb`] cache to the test.
+    #[cfg(feature = "unstable")]
+    pub(crate) fn with_fs_block_cache(self) -> TestBuilder<FsBlockCache> {
+        TestBuilder {
+            cache: FsBlockCache::new(),
+            seed: self.seed,
+            with_test_account: self.with_test_account,
+        }
+    }
+}
+
+impl<Cache> TestBuilder<Cache> {
+    /// Gives the test knowledge of the wallet seed for initialization.
+    pub(crate) fn with_seed(mut self, seed: SecretVec<u8>) -> Self {
+        // TODO remove
+        self.seed = Some(seed);
+        self
+    }
```

Comment:
I have a feeling that this can be removed, but I included it because some tests didn't provide a seed for `init_wallet_db` while others explicitly provided a `vec![]`, and I don't remember why the latter are doing that (there's probably some migration-related reason, though that seems odd for blank databases).

Reply:
Account initialization will also change with the `AccountBirthday` addition in #915, so it may be able to be removed after that.